File: FretTrainerEZ/Models/GuitarTuning.swift
import Foundation

struct GuitarTuning: Identifiable, Hashable {
    let id: String
    let name: String
    /// Open note for each string, index 0 = lowest string (string 6 in guitar convention)
    let strings: [Note]
    /// When true, display enharmonic notes as flats (e.g. Eb instead of D#)
    let useFlats: Bool

    var stringCount: Int { strings.count }

    // MARK: - Presets (top 10 tunings)

    static let standard     = GuitarTuning(id: "standard",    name: "Standard",       strings: [.E,  .A,  .D,  .G,  .B,  .E],  useFlats: false)
    static let dropD        = GuitarTuning(id: "dropD",       name: "Drop D",         strings: [.D,  .A,  .D,  .G,  .B,  .E],  useFlats: false)
    static let openG        = GuitarTuning(id: "openG",       name: "Open G",         strings: [.D,  .G,  .D,  .G,  .B,  .D],  useFlats: false)
    static let openD        = GuitarTuning(id: "openD",       name: "Open D",         strings: [.D,  .A,  .D,  .Fs, .A,  .D],  useFlats: false)
    static let dadgad       = GuitarTuning(id: "dadgad",      name: "DADGAD",         strings: [.D,  .A,  .D,  .G,  .A,  .D],  useFlats: false)
    static let openE        = GuitarTuning(id: "openE",       name: "Open E",         strings: [.E,  .B,  .E,  .Gs, .B,  .E],  useFlats: false)
    static let openA        = GuitarTuning(id: "openA",       name: "Open A",         strings: [.E,  .A,  .E,  .A,  .Cs, .E],  useFlats: false)
    static let halfStepDown = GuitarTuning(id: "halfStepDown",name: "Half Step Down", strings: [.Ds, .Gs, .Cs, .Fs, .As, .Ds], useFlats: true)
    static let fullStepDown = GuitarTuning(id: "fullStepDown",name: "Full Step Down", strings: [.D,  .G,  .C,  .F,  .A,  .D],  useFlats: false)
    static let dropC        = GuitarTuning(id: "dropC",       name: "Drop C",         strings: [.C,  .G,  .C,  .F,  .A,  .D],  useFlats: false)

    static let all: [GuitarTuning] = [
        standard, dropD, openG, openD, dadgad,
        openE, openA, halfStepDown, fullStepDown, dropC
    ]
}

File: android/app/src/main/kotlin/com/dontfretaboutitai/frettrainerez/models/GuitarTuning.kt
package com.dontfretaboutitai.frettrainerez.models

data class GuitarTuning(
    val id: String,
    val name: String,
    val strings: List<Note>,   // index 0 = lowest string
    val useFlats: Boolean = false
) {
    val stringCount: Int get() = strings.size

    companion object {
        val standard     = GuitarTuning("standard",     "Standard",        listOf(Note.E,  Note.A,  Note.D,  Note.G,  Note.B,  Note.E))
        val dropD        = GuitarTuning("dropD",        "Drop D",          listOf(Note.D,  Note.A,  Note.D,  Note.G,  Note.B,  Note.E))
        val openG        = GuitarTuning("openG",        "Open G",          listOf(Note.D,  Note.G,  Note.D,  Note.G,  Note.B,  Note.D))
        val openD        = GuitarTuning("openD",        "Open D",          listOf(Note.D,  Note.A,  Note.D,  Note.Fs, Note.A,  Note.D))
        val dadgad       = GuitarTuning("dadgad",       "DADGAD",          listOf(Note.D,  Note.A,  Note.D,  Note.G,  Note.A,  Note.D))
        val openE        = GuitarTuning("openE",        "Open E",          listOf(Note.E,  Note.B,  Note.E,  Note.Gs, Note.B,  Note.E))
        val openA        = GuitarTuning("openA",        "Open A",          listOf(Note.E,  Note.A,  Note.E,  Note.A,  Note.Cs, Note.E))
        val halfStepDown = GuitarTuning("halfStepDown", "Half Step Down",  listOf(Note.Ds, Note.Gs, Note.Cs, Note.Fs, Note.As, Note.Ds), useFlats = true)
        val fullStepDown = GuitarTuning("fullStepDown", "Full Step Down",  listOf(Note.D,  Note.G,  Note.C,  Note.F,  Note.A,  Note.D))
        val dropC        = GuitarTuning("dropC",        "Drop C",          listOf(Note.C,  Note.G,  Note.C,  Note.F,  Note.A,  Note.D))

        val all = listOf(standard, dropD, openG, openD, dadgad, openE, openA, halfStepDown, fullStepDown, dropC)
    }
}
